python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py

```diff
@@ -241,40 +260,58 @@ def to_default_asset_events(
                     attached_node_resource_props
                 )
+                check_name = event_node_info["node_name"]
+
+                if (
+                    context
+                    and AssetCheckKey(asset_key=attached_node_asset_key, name=check_name)
+                    in context.selected_asset_check_keys
+                ):
+                    is_test_successful = node_status == TestStatus.Pass
+                    severity = (
+                        AssetCheckSeverity.WARN
+                        if node_status == TestStatus.Warn
+                        else AssetCheckSeverity.ERROR
+                    )
 
-                yield AssetCheckResult(
-                    passed=is_test_successful,
-                    asset_key=attached_node_asset_key,
-                    check_name=event_node_info["node_name"],
-                    metadata=metadata,
-                    severity=severity,
-                )
-            else:
-                for upstream_unique_id in upstream_unique_ids:
-                    upstream_resource_props: Dict[str, Any] = manifest["nodes"].get(
-                        upstream_unique_id
-                    ) or manifest["sources"].get(upstream_unique_id)
-                    upstream_asset_key = dagster_dbt_translator.get_asset_key(
-                        upstream_resource_props
+                    yield AssetCheckResult(
+                        passed=is_test_successful,
+                        asset_key=attached_node_asset_key,
+                        check_name=check_name,
+                        metadata=metadata,
+                        severity=severity,
                     )
 
-                    yield AssetObservation(
-                        asset_key=upstream_asset_key,
+                # dbt's default indirect selection (eager) will select relationship tests
+                # on unselected assets, if they're compared with a selected asset.
+                # This doesn't match Dagster's default check selection which is to only
+                # select checks on selected assets. When we use eager, we may receive
+                # unexpected test results so we log those as observations as if
+                # asset checks were disabled.
+                else:
+                    logger.info(
+                        f"dbt test '{check_name}' will be logged as an AssetObservation instead of "
+                        "an AssetCheckResult because the asset check was not selected."
+                    )
+                    yield from self._yield_observation_events_for_test(
+                        dagster_dbt_translator=dagster_dbt_translator,
+                        validated_manifest=manifest,
+                        upstream_unique_ids=upstream_unique_ids,
                         metadata=metadata,
                     )
+            else:
+                yield from self._yield_observation_events_for_test(
```

Comment:
Let's clearly document each of this branches -- there's a lot of conditional happening in this yield function now, so want to make sure everything is understood for the future